File: tests/syntax/supported/r/contract.R
# Comprehensive R fixture exercising the highlight and locals queries.
library(methods)
suppressMessages(library(utils))

DEFAULTS <- list(min = 0L, max = 100L, label = "range")

#' Compute the euclidean norm of a numeric vector.
norm2 <- function(v) {
  sqrt(sum(v^2))
}

scale_values <- function(x, lower = 0, upper = 1) {
  rng <- range(x, na.rm = TRUE)
  if (diff(rng) == 0) {
    return(rep(lower, length(x)))
  }
  lower + (x - rng[1]) / diff(rng) * (upper - lower)
}

Point <- setRefClass("Point",
  fields = list(x = "numeric", y = "numeric"),
  methods = list(
    distance = function(other) {
      sqrt((x - other$x)^2 + (y - other$y)^2)
    }
  )
)

process <- function(items, ...) {
  items |>
    Filter(f = function(v) v %% 2 == 0) |>
    vapply(FUN = function(v) v * v, FUN.VALUE = numeric(1)) |>
    sum()
}

main <- function() {
  values <- c(1, 2, 3, 4, 5, NA)
  clean <- values[!is.na(values)]
  total <- 0
  i <- 1
  while (i <= length(clean)) {
    total <- total + clean[[i]]
    i <- i + 1
  }
  cat(sprintf("total=%d mean=%.2f\n", total, base::mean(clean)))
  repeat {
    if (total <= 0) break
    total <- total - 10
    if (total < 0) next
  }
  flags <- c(read = TRUE, write = FALSE)
  invisible(flags)
}

x = 10
20 -> y
z <<- x + y
main()

File: tests/syntax/supported/clojure/contract.clj
;; Comprehensive Clojure fixture exercising the highlight query.
(ns example.contract
  (:require [clojure.string :as str]
            [clojure.set :as set])
  (:import (java.time Instant)))

(def ^:private version "1.2.3")
(def limits {:min 0 :max 100})
(declare process)

(defmacro unless [test & body]
  `(if (not ~test)
     (do ~@body)))

(defrecord Point [x y])

(defn distance
  "Euclidean distance between two points."
  [{x1 :x y1 :y} {x2 :x y2 :y}]
  (let [dx (- x2 x1)
        dy (- y2 y1)]
    (Math/sqrt (+ (* dx dx) (* dy dy)))))

(defmulti describe :kind)
(defmethod describe :circle [shape]
  (str "circle r=" (:radius shape)))
(defmethod describe :default [_]
  "unknown")

(defn process [items]
  (->> items
       (filter even?)
       (map #(* % %))
       (reduce + 0)))

(defn -main [& args]
  (doseq [n (range 5)]
    (when-not (neg? n)
      (println (format "n=%d sq=%d" n (* n n)))))
  (let [chars [\a \b \c]
        flags #{:read :write}]
    (cond-> {:ok true}
      (seq args) (assoc :args (vec args))
      (contains? flags :read) (assoc :readable true))))

(comment
  (process [1 2 3 4 5 6])
  (distance (->Point 0 0) (->Point 3 4)))
